### ISSUE TITLE
Increasing creation time for new types/clusters, if an index is defined

### ISSUE THREAD

**Issue by benjaminbillet** [orientechnologies/orientdb#4617]
In the context of a converter from SQL to graph database, I benchmarked the creation of new types in OrientDB. Strangely, when an index is defined, the time spent for creating new types/clusters increases linearly (approximately) over time.

My benchmark performs the following steps, using the Java API with OrientDB 2.1rc4:
```
add 2 string properties to V (someProperty1, someProperty2)
for 1 to 10 do
    insert 10 new types
    insert 10000 vertices
endfor
```
Concretely, in each iteration I measure only the time spent to create the 10 new types, in three different case:
- no index
- one index defined for someProperty1
- one index defined for someProperty1 and one index defined for someProperty2

It appears that the time spent for creating new types is constant if no indices are defined. However, with an index, the creation time increases linearly to the number of existing types. If we increase the number of indices, the creation time seems to increase linearly as well.
![Benchmark](http://benjaminbillet.fr/media/orientdb_cluster_benchmark.png)

The problem occurs with SB-Tree and Hash indices, unique or not unique. It occurs as well when the clusters are empty (no vertex insertion). Intuitively, I suspect a problem related to index allocation or index rebalacing.

Here is the quick&dirty code I runned for the benchmark: http://benjaminbillet.fr/media/OrientDBDatabaseBenchmarks.java

**Comment by compensator** [comment]
for(int i = 0; i < nbSteps; i++)
		{						
			long startTime = System.nanoTime();
			for(int j = nbTypes; j < goal; j++)
			{
				graph.createVertexType("MyClass" + j);
			}
			long delta = System.nanoTime() - startTime;
			totalTypeCreationTime += delta;
			System.out.println("create type: " + nbTypes + " -> " + goal + " in " + ((double) delta / NANO) + "s");
			
			if(nbVertices > 0)
			{
				startTime = System.nanoTime();
				for(int j = 0; j < nbVertices; j++)
				{
					graph.addVertex("class:MyClass" + (j % goal), "someProperty1", "value" + j, "someProperty2", "value" + (i * nbVertices + j));
				}
				delta = System.nanoTime() - startTime;
				totalInsertionTime += delta;
				System.out.println("insert vertices: " + (nbVertices * nbTypes) + " -> " + (nbVertices * goal) + " in " + ((double) delta / NANO) + "s");
			}
			
			nbTypes += nbTypesPerSteps;
			goal += nbTypesPerSteps;
		}

It appears that your "for" loop is not assigning the new nbTypes value to the loop control variable "i", so the loop limit is increasing by 10 with each iteration...10, 20, 30, 40...

**Comment by benjaminbillet** [comment]
I don't follow you. The variable `nbTypes` is not used in the first "for" loop. If we simplify my code, we have:
```java
int nbSteps = 10;
int nbTypesPerSteps = 10;
int nbTypes = 0;
int goal = 10;
for(int i = 0; i < nbSteps; i++) // nbSteps is never increased
{						
	for(int j = nbTypes; j < goal; j++)
	{
		graph.createVertexType("MyClass" + j);
	}
	nbTypes += nbTypesPerSteps;
	goal += nbTypesPerSteps;
}
```
The first "for" will iterate from 0 to `nbSteps` (10). The second "for" loop will iterate from `nbTypes` to `goal`, which are initially set to 0 and 10, and then increased by `nbTypesPerSteps` (10) at the end of each iteration of the first "for" loop.

Anyway, I created this benchmark because I first observed this type creation time behavior in the project I'm currently working on.

**Comment by laa** [comment]
Hi,
We can mitigate this overhead for DBs of average size (in short terms) but not for big database. Reason of this overhead is that indexes are used random IO reads which are much slower than RAM reads. As result we have linear complexity during insertion because more and more pages are on disk not in cache.

**Comment by laa** [comment]
But we have long term strategy for big databases too :-)

**Comment by laa** [comment]
Oh I see sorry it is not a bug )) We do forcesynch when we create new types the more data you have more data to flush. so no surprise for me.

**Comment by smolinari** [comment]
How many vertices are created once the test is finished? 

Scott

**Comment by benjaminbillet** [comment]
@laa I don't understand why it is not a bug, because with no data (empty database) the problem is the same. In any case, this problem should be mitigated, as creating many types with many indices is not an exotic use case. I work on a project that creates a lot of types and, after some time, creating a new type takes thirty minutes... Currently, my workaround is to use a straightforward in-memory indexing instead of OrientDB indexing, given that I have less than 200000 vertices. However, this would not be suitable for large databases.

Another workaround is to create the indices after all the types are created. However, this is a "static" solution; if I want to create new types after that, the creation time will be huge.

@smolinari I performed my benchmarks first with only 100000 vertices and then with an empty database: I got nearly the same creation times. If you want to see it by yourself, you can run my benchmarking code and set the number of vertices to 0.

**Comment by lvca** [comment]
I agree we should fix this. Not as a bug, but a "performance" issue.

**Comment by laa** [comment]
Problem of this bug is that we require to flush disk cache in case of class\cluster\index creation. Recently we decided to postpone change which allows to prevent force flush of data which in turn will fix this bug. 
So I postpone this change to next version.

**Comment by laa** [comment]
Fixed in 2.2 version
